spec/services/declarations/create_spec.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Declarations::Create, type: :model do
+  let(:lead_provider) { LeadProvider.all.sample }
+  let(:cohort) { create(:cohort, :current) }
+  let(:course_group) { CourseGroup.find_by(name: "leadership") || create(:course_group, name: "leadership") }
+  let(:course) { create(:course, :sl, course_group:) }
+  let!(:schedule) { create(:schedule, :npq_leadership_autumn, course_group:, cohort:) }
+  let(:application) { create(:application, :accepted, cohort:, course:, lead_provider:) }
+  let(:participant) { application.user }
+  let(:participant_id) { participant.ecf_id }
+  let(:declaration_type) { "started" }
+  let(:declaration_date) { schedule.applies_from + 1.day }
+  let(:course_identifier) { course.identifier }
+  let(:has_passed) { true }
+  let(:params) do
+    {
+      lead_provider:,
+      participant_id:,
+      declaration_type:,
+      declaration_date: declaration_date.rfc3339,
+      course_identifier:,
+      has_passed:,
+    }
+  end
+
+  before { create(:contract, course:, cohort:, lead_provider:) }
+
+  subject(:service) { described_class.new(**params) }
+
+  describe "validations" do
+    it { is_expected.to validate_presence_of(:lead_provider) }
+    it { is_expected.to validate_presence_of(:participant_id).with_message("The property '#/participant_id' must be present") }
```

Comment:
Can we avoid hard coding the error messages

Reply:
There are still a few hard-coded translations kicking about; you can test the ones with a variable as well using the key like this:

```
    record
      .errors
      .add(:participant_id, :declaration_must_be_before_withdrawal_date, withdrawal_date: latest_state.created_at.rfc3339)
```

```
expect(subject.errors.first).to have_attributes(attribute: :participant_id, type: :declaration_must_be_before_withdrawal_date, options: { withdrawal_date: application.application_states.last.created_at.rfc3339 })

```